crypto/rsa/rsa_sp800_56b_gen.c

```diff
@@ -209,6 +209,131 @@ static int rsa_validate_rng_strength(EVP_RAND_CTX *rng, int nbits)
     return 1;
 }
 
+
+DEFINE_STACK_OF(BIGNUM)
+
+/*
+ *
+ * Using p & q, calculate dmp1, dmq1, iqmp
+ * Placing them on their respective BIGNUM stacks
+ *
+ * See SP800-56Br1
+ *   6.3.1.1 rsakpg1 - basic (Steps 3-4)
+ *   6.3.1.3 rsakpg1 - crt   (Step 5)
+ *
+ * Params:
+ *  p,q - input primes
+ *  d,n - input key values
+ *  e - input modulus
+ *  nbits - requested size of our computed values
+ *  exps, coeffs - output list of computed values for dmp1, dmq1, iqmp
+ *  modify - boolean value to indicate if the provided inputs should be modified in place
+ *           see use in ossl_rsa_sp800_56b_derive_params_from_pq
+ *
+ * Returns: -1 = error,
+ *           0 = d is too small,
+ *           1 = success.
+ *
+ */
+int ossl_rsa_sp800_56b_build_exp_coeffs_from_pq(BIGNUM *p, BIGNUM *q, BIGNUM *d, BIGNUM *n, int nbits,
+                                             const BIGNUM *e, STACK_OF(BIGNUM) *exps, STACK_OF(BIGNUM) *coeffs, BN_CTX *ctx, int modify)
+{
+    int ret = -1;
+    BIGNUM *p1, *q1, *lcm, *p1q1, *gcd;
```

Comment:
Instead of this refactoring I would keep this inside the ossl_rsa_sp800_56b_derive_params_from_pq() function and just skipped the n and d generation based on the e parameter being NULL.

That would make the change here much smaller.

Reply:
but e is never null in this call path.  If an e value is not provided in the keygen path, its set to a default value of 65537, and in the fromdata path we validate that we have an e parameter, or we return an error.

Reply:
ossl_rsa_sp800_56b_derive_params_from_pq() has an `e` parameter. But it also has a `rsa` parameter which can contain `e`. The fips keygen would pass the e value as the parameter as it currently does where you would pass NULL there, indicating that you do not want to touch n, d, e values in the rsa object and just update the exponents, and coeffs.